[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76081

Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

dzdang committed:
[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op

Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op"


Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).

dzdang commented:
land the max pool 2d PR first: https://github.com/pytorch/pytorch/pull/76081
and then come back and fix the macros here

dzdang commented:
rebase on master after https://github.com/pytorch/pytorch/pull/76129 is merged with master. should fix the current CI errors

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op"


Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op"


Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op"


Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op"


Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Pooling.cpp:
```diff
@@ -14,6 +14,8 @@
 #include <ATen/ATen.h>
 #include <ATen/native/Pool.h>
 #include <ATen/native/TensorIterator.h>
+#include <ATen/quantized/Quantizer.h>
```
why are we importing quantizer

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Pooling.cpp:
```diff
@@ -41,6 +43,31 @@ void check_maxpool2d_params(
+#ifdef USE_CUDA
+// #if AT_CUDNN_ENABLED()
+// #if HAS_CUDNN_V8()
+    // TODO: limit this to per channel quantized tensors for now, though should be easy to adapt
```
per channel --> per tensor?

jerryzh168 replied:
per channel --> per tensor?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Pooling.cpp:
```diff
@@ -41,6 +43,31 @@ void check_maxpool2d_params(
+// #if HAS_CUDNN_V8()
+    // TODO: limit this to per channel quantized tensors for now, though should be easy to adapt
+    // to per channel quantized tensors
+    TORCH_CHECK(input.qscheme() == at::kPerTensorAffine, "adaptive_avg_pool2d_quantized_cuda oonly supports per channel quantized tensors");
```
same here

jerryzh168 replied:
same here

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized cuda adaptive average pool2d op"


Summary:
The current implementation of quantized cuda adaptive average pooling uses the following:
dequant -> fp32 adaptive average pooling -> quant. This is the same numerically as quantized adaptive average pooling. This is not the ideal implementation, as we desire to operate on the quantized values directly. However, we are currently blocked on this as we are waiting for cudnn's 8.5.0 release, which is anticipated to support adaptive average pooling. When that support is made available, we will use it directly.

Test plan:
```
python test/test_quantization.py TestQuantizedOps.test_adaptive_avg_pool
```

Differential Revision: [D35768751](https://our.internmc.facebook.com/intern/diff/D35768751)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35768751).
